fstar-helpers/minicore/src/core_arch/x86/interpretations.rs

```diff
@@ -0,0 +1,183 @@
+        // let hey (#t:inttype) (#t':inttype) (x:int_t t): Lemma (Rust_primitives.cast x == cast t t' x) = admit ()
+
+        let ${flatten_circuit} (): FStar.Tactics.Tac unit =
+            let open Tactics.Circuits in
+            flatten_circuit
+                [
+                    "Minicore";
+                    "FStar.FunctionalExtensionality";
+                    `%Rust_primitives.cast_tc; `%Rust_primitives.unsize_tc;
+                    "Core.Ops"; `%(.[]);
+                ]
+                (top_levels_of_attr (` $LIFT_LEMMA ))
+                (top_levels_of_attr (` $SIMPLIFICATION_LEMMA ))
+                (top_levels_of_attr (` $ETA_MATCH_EXPAND ));
+            ()
+        "#
+        )]
+        /// F* tactic: specialization of `Tactics.Circuits.flatten_circuit`.
+        pub fn flatten_circuit() {}
+    }
+
+    #[cfg(all(test, any(target_arch = "x86", target_arch = "x86_64")))]
+    mod tests {
+        use crate::abstractions::bitvec::BitVec;
+        use crate::core_arch::x86::upstream;
+
+        /// Helper trait to generate random values
+        pub trait HasRandom {
+            fn random() -> Self;
+        }
+
+        impl HasRandom for i32 {
+            fn random() -> Self {
+                use rand::prelude::*;
+                let mut rng = rand::rng();
+                rng.random()
+            }
+        }
+
+        impl<const N: u64> HasRandom for BitVec<N> {
+            fn random() -> Self {
+                BitVec::rand()
+            }
+        }
+
+        /// Derives tests for a given intrinsics. Test that a given intrisics and its model compute the same thing over random values (1000 by default).
+        macro_rules! mk {
+            ($([$N:literal])?$name:ident$({$(<$($c:literal),*>),*})?($($x:ident : $ty:ident),*)) => {
+                #[test]
+                fn $name() {
+                    #[allow(unused)]
+                    const N: usize = {
+                        let n: usize = 1000;
+                        $(let n: usize = $N;)?
+                        n
+                    };
+                    mk!(@[N]$name$($(<$($c),*>)*)?($($x : $ty),*));
+                }
+            };
+            (@[$N:ident]$name:ident$(<$($c:literal),*>)?($($x:ident : $ty:ident),*)) => {
+                for _ in 0..$N {
+                    $(let $x = $ty::random();)*
+                    assert_eq!(super::$name$(::<$($c,)*>)?($($x.into(),)*), unsafe {
+                        BitVec::from(upstream::$name$(::<$($c,)*>)?($($x.into(),)*)).into()
+                    });
+                }
+            };
+            (@[$N:ident]$name:ident<$($c1:literal),*>$(<$($c:literal),*>)*($($x:ident : $ty:ident),*)) => {
+                let one = || {
+                    mk!(@[$N]$name<$($c1),*>($($x : $ty),*));
+                };
+                one();
+                mk!(@[$N]$name$(<$($c),*>)*($($x : $ty),*));
+            }
+        }
+
+        mk!(_mm256_set1_epi32(x: i32));
```

Comment:
How comprehensive are these tests? Do they cover all the intrinsics you define in minicore?

Reply:
Can you just leave a comment, maybe in this file or in a readme, instructing people on the steps they should take to model new intrinsics. E.g. add a model to int-vec, link with interpretations, add a test etc?

Reply:
Sounds good! 👍

Reply:
Done, I added some doc comments on `x86.rs`, here is a capture of the rendered version with `cargo doc`:
<img width="478" alt="Capture d’écran 2025-04-30 à 17 10 56" src="https://github.com/user-attachments/assets/bd596b1d-e96e-45e3-94bf-266490edcb76" />